src/jvm/main/org/jetbrains/kotlinx/lincheck/util/Logger.kt

```diff
@@ -0,0 +1,91 @@
+/*
+ * Lincheck
+ *
+ * Copyright (C) 2019 - 2025 JetBrains s.r.o.
+ *
+ * This Source Code Form is subject to the terms of the
+ * Mozilla Public License, v. 2.0. If a copy of the MPL was not distributed
+ * with this file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package org.jetbrains.kotlinx.lincheck.util
+
+import java.io.*
+import java.text.SimpleDateFormat
+import java.util.*
+
+object LincheckLogger {
+    private val LINE_SEPARATOR = System.lineSeparator()
+    private val formatter: ThreadLocal<SimpleDateFormat> = object : ThreadLocal<SimpleDateFormat>() {
+        override fun initialValue(): SimpleDateFormat {
+            return SimpleDateFormat("dd.MM.yyyy HH:mm:ss")
+        }
+    }
+
+    private val logFilename = System.getProperty("lincheck.log.file") ?: null
+    private val logFile = if (logFilename != null) File(logFilename) else null
+    private var logWriter: Writer? = null
+
+    init {
+        println("property: lincheck.log.file = " + System.getProperty("lincheck.log.file"))
+        logFile?.let {
+            initFile(it)
+            logWriter = BufferedWriter(FileWriter(it))
+            println("Lincheck log file could be found at: ${it.absolutePath}")
+        }
+    }
+
+    /**
+     * Logs will be printed to log file if it exists.
+     */
+    fun log(s: String) {
```

Comment:
Please implement the logging as follows:

```suggestion
    inline fun log(logLevel: LogLevel, lazyMessage: () -> String) { 
        if (logLevel <= this.logLevel) {
            write(message)
        }
    }
```

By making the function `inline` and accepting the lambda to produce the message, you avoid actually allocating memory and creating `String` object in case when the log message will not be printed because of the insufficient logging level:

https://kotlinlang.org/docs/inline-functions.html